stdlib/src/collections/deque.mojo

```diff
@@ -0,0 +1,313 @@
+        self.head = existing.head
+        self.tail = existing.tail
+        self.minlen = existing.minlen
+        self.maxlen = existing.maxlen
+        self.shrink = existing.shrink
+
+    fn __del__(owned self):
+        """Destroys all elements in the deque and free its memory."""
+        for i in range(len(self)):
+            offset = (self.head + i) & (self.capacity - 1)
+            (self.data + offset).destroy_pointee()
+        self.data.free()
+
+    # ===-------------------------------------------------------------------===#
+    # Trait implementations
+    # ===-------------------------------------------------------------------===#
+
+    @always_inline
+    fn __bool__(self) -> Bool:
+        """Checks whether the deque has any elements or not.
+
+        Returns:
+            `False` if the deque is empty, `True` if there is at least one element.
+        """
+        return self.head != self.tail
+
+    @always_inline
+    fn __len__(self) -> Int:
+        """Gets the number of elements in the deque.
+
+        Returns:
+            The number of elements in the deque.
+        """
+        return (self.tail - self.head) & (self.capacity - 1)
+
+    fn __getitem__(ref [_]self, idx: Int) -> ref [self] ElementType:
+        """Gets the deque element at the given index.
+
+        Args:
+            idx: The index of the element.
+
+        Returns:
+            A reference to the element at the given index.
+        """
+        normalized_idx = idx
+
+        debug_assert(
+            -len(self) <= normalized_idx < len(self),
+            "index: ",
+            normalized_idx,
+            " is out of bounds for `Deque` of size: ",
+            len(self),
+        )
+
+        if normalized_idx < 0:
+            normalized_idx += len(self)
+
+        offset = (self.head + normalized_idx) & (self.capacity - 1)
+        return (self.data + offset)[]
+
+    # ===-------------------------------------------------------------------===#
+    # Methods
+    # ===-------------------------------------------------------------------===#
+
+    fn append(inout self, owned value: ElementType):
+        """Appends a value to the right side of the deque.
+
+        Args:
+            value: The value to append.
+        """
+        if len(self) == self.maxlen:
+            (self.data + self.head).destroy_pointee()
+            self.head = (self.head + 1) & (self.capacity - 1)
+
+        (self.data + self.tail).init_pointee_move(value^)
+        self.tail = (self.tail + 1) & (self.capacity - 1)
+
+        if self.head == self.tail:
+            self._realloc(self.capacity << 1)
+
+    fn extend(inout self, owned values: List[ElementType]):
+        """Extends the right side of the deque by consuming elements of the list argument.
+
+        Args:
+            values: List whose elements will be added at the right side of the deque.
+        """
+        for value in values:
+            self.append(value[])
```

Comment:
I will have to think more about this. Perhaps it makes sense to go for the most efficient implementation of `extend` precalculating the final capacity, checking if re-allocation will be needed or not, accounting for the `maxlen` effect etc., and then preallocate buffer and `move_pointee_into()` only the final elements.

Reply:
Refactored that. From 2 lines of code to 42 :)

My naive testing shows me speed improvements from 15% (in "standard" situations) to about 4x time improvement when we try to extend the deque having a small `maxlen` with the huge list.

But I do not like how we get coupled with the internal implementation of the List type. If we refactor List to use `_data` instead of `data` (desired, probably), Deque implementation would break. Can we achieve the same effect using only the public API of the List? Like doing it in a way we work with the VariadicListMem:
```
src = UnsafePointer.address_of(values[i])
src.move_pointee_into(dst)
```

On the other side, we are still going into the guts of VariadicListMem with `variadic_list._is_owned = False` (same way as the List implementation does) and with the `values.size = 0`. So, I will leave you to decide how much coupling do you tolerate, if at all.